File: runtime/org.eclipse.etrice.runtime.cpp/src/common/messaging/MessageService.cpp
/*******************************************************************************
 * Copyright (c) 2012 Draeger Medical GmbH (http://www.draeger.com).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * CONTRIBUTORS:
 * 		Peter Karlitschek (initial contribution)
 *
 *******************************************************************************/

#include "common/messaging/MessageSeQueue.h"
#include "common/messaging/MessageService.h"
#include "common/messaging/MessageServiceController.h"
#include "common/messaging/RTServices.h"
#include "osal/etTime.h"
#include <new>

namespace etRuntime {


MessageService::MessageService(IRTObject* parent, IMessageService::ExecMode mode, int node, int thread, const String& name,
		IMessageMemory* memory, int priority) :
		RTObject(parent, name),
		m_running(false),
		m_execMode(mode),
		m_lastMessageTimestamp(0),
		m_address(node, thread, 0),
		m_messageQueue(this, "Queue"),
		m_messageDispatcher(this, m_address.createInc(), "Dispatcher"),
		m_messageMemory(memory) {

	etTime interval;
	interval.sec = 0;
	interval.nSec = 0;
	MessageService_init(interval, priority);
}

MessageService::MessageService(IRTObject* parent, IMessageService::ExecMode mode, etTime interval, int node, int thread,
		const String& name, IMessageMemory* memory, int priority) :
		RTObject(parent, name),
		m_running(false),
		m_execMode(mode),
		m_lastMessageTimestamp(0),
		m_address(node, thread, 0),
		m_messageQueue(this, "Queue"),
		m_messageDispatcher(this, m_address.createInc(), "Dispatcher"),
		m_messageMemory(memory) {

	MessageService_init(interval, priority);
}

void MessageService::MessageService_init(etTime interval, int priority) {

	/* init mutexes and semaphores */
	etMutex_construct(&m_mutex);
	etSema_construct(&m_executionSemaphore);

	/* init thread */
	etThread_construct(&m_thread, static_cast<etStacksize>(1024), static_cast<etPriority>(priority), (etThreadname)"MessageService",
			MessageService::run, static_cast<void *>(this));

	// check and set priority
//			assert priority >= Thread.MIN_PRIORITY : ("priority smaller than Thread.MIN_PRIORITY (1)");
//			assert priority <= Thread.MAX_PRIORITY : ("priority bigger than Thread.MAX_PRIORITY (10)");
//			this.setPriority(priority);

	if (m_execMode == IMessageService::POLLED || m_execMode == IMessageService::MIXED) {
		/* init timer */
		etTimer_construct(&m_timer, &interval, MessageService::pollingTask, static_cast<void *>(this));
	}
}

MessageService::~MessageService() {
	while(m_messageQueue.getSize() > 0) {
		const Message* msg = m_messageQueue.pop();
		msg->~Message();
		returnMessageBuffer(msg);
	}
	etMutex_destruct(&m_mutex);
	etSema_destruct(&m_executionSemaphore);
	etThread_destruct(&m_thread);
	if (m_execMode == IMessageService::POLLED || m_execMode == IMessageService::MIXED) {
		etTimer_destruct(&m_timer);
	}
	delete m_messageMemory;
}

void MessageService::start() {
	m_running = true;
	etThread_start(&m_thread);
	if (m_execMode == IMessageService::POLLED || m_execMode == IMessageService::MIXED) {
		etTimer_start(&m_timer);
	}
}

void MessageService::run() {
	while (m_running) {
		etMutex_enter(&m_mutex);
		const Message* msg = m_messageQueue.pop(); // get next Message from Queue
		etMutex_leave(&m_mutex);
		if (msg == 0) {
			// no message in queue -> wait till Thread is notified
			etSema_waitForWakeup(&m_executionSemaphore);
		} else {
			//TODO: set timestamp
			// m_lastMessageTimestamp = System.currentTimeMillis();
			m_messageDispatcher.receive(msg);
		}
	}

	RTServices::getInstance().getMsgSvcCtrl().setMsgSvcTerminated(*this);
}

void MessageService::receive(const Message* msg) {
	etMutex_enter(&m_mutex);
	if (msg != 0) {
		m_messageQueue.push(const_cast<Message*>(msg));
	}
	etSema_wakeup(&m_executionSemaphore);
	etMutex_leave(&m_mutex);
}

Address MessageService::getFreeAddress() {
	etMutex_enter(&m_mutex);
	Address address = m_messageDispatcher.getFreeAddress();
	etMutex_leave(&m_mutex);

	return address;
}

void MessageService::freeAddress(const Address& addr) {
	etMutex_enter(&m_mutex);
	m_messageDispatcher.freeAddress(addr);
	etMutex_leave(&m_mutex);
}

void MessageService::addMessageReceiver(IMessageReceiver& receiver) {
	etMutex_enter(&m_mutex);
	m_messageDispatcher.addMessageReceiver(receiver);
	etMutex_leave(&m_mutex);
}

void MessageService::removeMessageReceiver(IMessageReceiver& receiver) {
	etMutex_enter(&m_mutex);
	m_messageDispatcher.removeMessageReceiver(receiver);
	etMutex_leave(&m_mutex);
}

void MessageService::addPollingMessageReceiver(IMessageReceiver& receiver) {
	etMutex_enter(&m_mutex);
	m_messageDispatcher.addPollingMessageReceiver(receiver);
	etMutex_leave(&m_mutex);
}

void MessageService::removePollingMessageReceiver(IMessageReceiver& receiver) {
	etMutex_enter(&m_mutex);
	m_messageDispatcher.removePollingMessageReceiver(receiver);
	etMutex_leave(&m_mutex);
}

Message* MessageService::getMessageBuffer(int size) {
	etMutex_enter(&m_mutex);
	Message* buffer = m_messageMemory->getMessageBuffer(size);
	etMutex_leave(&m_mutex);
	return buffer;
}

void MessageService::returnMessageBuffer(const Message* buffer) {
	etMutex_enter(&m_mutex);
	m_messageMemory->returnMessageBuffer(buffer);
	etMutex_leave(&m_mutex);
}

String MessageService::toString() const {
	return getName() + " " + getAddress().toID().c_str();
}

void MessageService::terminate() {
	if (m_execMode == IMessageService::POLLED || m_execMode == IMessageService::MIXED) {
		etTimer_stop(&m_timer);
	}
	if (m_running) {
		m_running = false;
		etSema_wakeup(&m_executionSemaphore);
	}

}

// called by osal timer, thread ?
void MessageService::pollingTask() {
	if (m_running) {
		Message* pollingMessage = getMessageBuffer(sizeof(Message));
		if (pollingMessage) {
			new (pollingMessage) Message(m_messageDispatcher.getAddress(), 0);
			receive(pollingMessage);
		}
		else {
			// TODO JB: error handling for pollingMessage == NULL
		}
	}
}

} /* namespace etRuntime */

File: runtime/org.eclipse.etrice.runtime.java/src/org/eclipse/etrice/runtime/java/messaging/MessageService.java
/*******************************************************************************
 * Copyright (c) 2010 protos software gmbh (http://www.protos.de).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * 
 * CONTRIBUTORS:
 * 		Thomas Schuetz and Henrik Rentz-Reichert (initial contribution)
 * 
 *******************************************************************************/


package org.eclipse.etrice.runtime.java.messaging;


/**
 * The MessageService is the backbone of the asynchronous communication inside a SubSystem
 * It usually contains a thread a message queue and a dispatcher
 * 
 * @author Thomas Schuetz (initial contribution)
 * @author Henrik Rentz-Reichert (extending RTObject, implementing Runnable)
 *
 */
public class MessageService extends AbstractMessageService {

	private boolean running = false;
	
	private Thread thread;
	private int priority;
	private long lastMessageTimestamp;

	public MessageService(IRTObject parent, ExecMode mode, int node, int thread, String name) {
		this(parent, mode, 0, node, thread, name, Thread.NORM_PRIORITY);
	}
	
	public MessageService(IRTObject parent, ExecMode mode, int nsec, int node, int thread, String name) {
		this(parent, mode, nsec, node, thread, name, Thread.NORM_PRIORITY);
	}
	
	public MessageService(IRTObject parent, ExecMode mode, int nsec, int node, int thread, String name, int priority) {
		super(parent, "MessageService_"+name, node, thread);
		
		this.priority = priority;

		assert priority >= Thread.MIN_PRIORITY : ("priority smaller than Thread.MIN_PRIORITY (1)"); 
		assert priority <= Thread.MAX_PRIORITY : ("priority bigger than Thread.MAX_PRIORITY (10)"); 
	}

	public void run() {
		running = true;
		
		while (running) {
			Message msg = null;
			
			// get next Message from Queue
			synchronized(this) {
				msg = getMessageQueue().pop();
			}
			
			if (msg == null) {
				// no message in queue -> wait until Thread is notified
				try {
					synchronized(this) {
						if (!running)
							return;
						wait();
					}
				}
				catch (InterruptedException e) {
				}
			}
			else {
				// process message
				lastMessageTimestamp = System.currentTimeMillis();
				getMessageDispatcher().receive(msg);
			}
		}
	}

	/* (non-Javadoc)
	 * @see org.eclipse.etrice.runtime.java.messaging.AbstractMessageService#receive(org.eclipse.etrice.runtime.java.messaging.Message)
	 */
	@Override
	public synchronized void receive(Message msg) {
		super.receive(msg);
		
		// wake up thread to process message
		notifyAll();
	}
	
	/* (non-Javadoc)
	 * @see org.eclipse.etrice.runtime.java.messaging.AbstractMessageService#getFreeAddress()
	 */
	@Override
	public synchronized Address getFreeAddress() {
		return super.getFreeAddress();
	}
	
	/* (non-Javadoc)
	 * @see org.eclipse.etrice.runtime.java.messaging.AbstractMessageService#addMessageReceiver(org.eclipse.etrice.runtime.java.messaging.IMessageReceiver)
	 */
	@Override
	public synchronized void addMessageReceiver(IMessageReceiver receiver) {
		super.addMessageReceiver(receiver);
	}

	/* (non-Javadoc)
	 * @see org.eclipse.etrice.runtime.java.messaging.AbstractMessageService#removeMessageReceiver(org.eclipse.etrice.runtime.java.messaging.IMessageReceiver)
	 */
	@Override
	public synchronized void removeMessageReceiver(IMessageReceiver receiver) {
		super.removeMessageReceiver(receiver);
	}
	
	/* (non-Javadoc)
	 * @see org.eclipse.etrice.runtime.java.messaging.AbstractMessageService#freeAddress(org.eclipse.etrice.runtime.java.messaging.Address)
	 */
	@Override
	public synchronized void freeAddress(Address addr) {
		super.freeAddress(addr);
	}
	
	public synchronized void terminate() {
		if (running) {
			running = false;
			notifyAll();
		}
	}

	/* (non-Javadoc)
	 * @see org.eclipse.etrice.runtime.java.messaging.IMessageService#setThread(java.lang.Thread)
	 */
	public void setThread(Thread thread) {
		this.thread = thread;
		
		thread.setPriority(priority);
	}

	/* (non-Javadoc)
	 * @see org.eclipse.etrice.runtime.java.messaging.IMessageService#getThread()
	 */
	public Thread getThread() {
		return thread;
	}

	protected long getLastMessageTimestamp() {
		return lastMessageTimestamp;
	}

}
